src/main/scala/chisel3/util/Queue.scala

```diff
@@ -185,6 +188,87 @@ object Queue {
     }
   }
 
+  /** A factory for creating shadow queues.  This is created using the
+    * `withShadowQueue` method.
+    */
+  class ShadowFactory[T <: Data] private[Queue] (
+    enq:            ReadyValidIO[T],
+    deq:            ReadyValidIO[T],
+    entries:        Int,
+    pipe:           Boolean,
+    flow:           Boolean,
+    useSyncReadMem: Boolean,
+    flush:          Option[Bool]) {
+
+    /** Create a "shadow" `Queue` in a specific layer that will be queued and
+      * dequeued in lockstep with an original `Queue`.  Connections are made
+      * using `BoringUtils.tapAndRead` which allows this method to be called
+      * anywhere in the hierarchy.
+      *
+      * An intended use case of this is as a building block of a "shadow" design
+      * verification datapath which augments an existing design datapath with
+      * additional information.  E.g., a shadow datapath that tracks transations
+      * in an interconnect.
+      *
+      * @param data a hardware data that should be enqueued together with the
+      * original `Queue`'s data
+      * @param layer the `Layer` in which this queue should be created
+      * @return a layer-colored `Valid` interface of probe type
+      */
+    def apply[A <: Data](data: A, layer: Layer): Valid[A] = {
+      val shadowDeq = Wire(Probe(Valid(chiselTypeOf(data)), layer))
+
+      block(layer) {
+        val shadowEnq = Wire(Decoupled(chiselTypeOf(data)))
+        shadowEnq.ready :<= true.B
```

Comment:
Should these just use the tapped ready and valid/ the ready from the shadow queue? Again is there any reason this ready would ever get out of sync with the original?

Reply:
I actually made a mistake here. This ready is entirely unused. Only the valid is used. The enqueue ready will be driven by the queue and then optimized away. This assumes that the two queues are going to work in lockstep.

While this _should_ always work in lockstep, this could get out of sync if there is clock gating or the queue and its shadow are on different clock domains. I've added an assert that the ready works in lockstep in : 424c92e3d